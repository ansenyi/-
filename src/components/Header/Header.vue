<template>
  <header>
    <div class="logo">
      <p>虹膜识别门禁管理系统</p>
    </div>
    <slot></slot>
    <div id="placeholder" v-if="$route.name === 'Login'"></div>
    <nav id="menu" v-else>
      <span>{{ adminName }}，欢迎你</span>
      <router-link to="/menu">主菜单</router-link>
      <router-link class="email" to="/message" v-show="!redShow"></router-link>
      <router-link class="email-red" to="/message" v-show="redShow"></router-link>
      <router-link to="/upload">软件上传</router-link>
      <router-link to="/history">历史记录</router-link> 
      <router-link to="/password">密码修改</router-link> 
      <!--<router-link to="" style="padding-right: 15px;">考勤管理</router-link>-->|
      <a class="logout" href="javascript:;" @click="logout()">退出</a>
      <!--<router-link to="/help">帮助</router-link>-->
    </nav>
  </header>
</template>

<script>
  import { _debounce } from '../../common/js/util'
  import Bus from '../../common/js/Bus'
  export default {
    name: 'header',
    data () {
      return {
        redShow: false,
        logout: _debounce(this.logoutHandler)
      }
    },
    computed: {
      adminName () {
        try {
          let name = window.localStorage.getItem('ADMIN_NAME')
          if (!name) {
            return null
          }
          return name
        } catch (error) {
          console.log('取name出错')
          console.log(error)
          return null
        }
      }
    },
    mounted () {
      if (this.$route.name !== 'Login') {
        // console.log('header mounted;on-hasMessage')
        Bus.$on('hasMessage', (show) => {
          this.redShow = show
        })
        Bus.$emit('requireMsg')
      }
    },
    methods: {
      logoutHandler () {
        this.$http.post('logout', '').then((response) => {
          console.log('退出时请求到的数据：')
          console.log(response)
          if (response.data.status) {
            console.log('退出成功！')
            this.clearToken()
            this.clearName()
            Bus.$emit('endMessageTimer')
            Bus.$emit('resetMsgData')
            this.$router.replace('/')
          } else {
            this.$emit('alertMsg', response.data.msg)
          }
        }).catch((error) => {
          console.log(error)
          this.$emit('alertMsg', '退出失败，请重试。')
        })
      },
      clearToken () {
        let localStorage = window.localStorage
        try {
//          if (localStorage.getItem('JWT_TOKEN') !== null) {
          if (localStorage.getItem(window.myToken) !== null) {
            localStorage.removeItem('JWT_TOKEN')
            console.log('logout token')
          }
        } catch (err) {
          console.log('logout token 出错')
          console.log(err)
        }
      },
      clearName () {
        let localStorage = window.localStorage
        try {
          if (localStorage.getItem('ADMIN_NAME') !== null) {
            localStorage.removeItem('ADMIN_NAME')
            console.log('logout name')
          }
        } catch (err) {
          console.log('logout name 出错')
          console.log(err)
        }
      }
    }
  }
</script>

<style lang="stylus" scoped>
  header
    position fixed
    top 0
    left 0
    z-index 997
    display flex
    width 100%
    height 40px
    background #333538
    .logo
      flex 1 1 50%
      height 40px
      vertical-align middle
      p
        width 300px
        text-align center
        color #fff
        font-size 20px
        margin-top 10px
        cursor default
    #placeholder
      flex 0 0 50%
    #menu
      flex 1 1 26%
      height 40px
      width 2100px !important
      line-height 40px
      font-size 14px
      text-align center
      overflow hidden
      color #fff
      /*background-color red*/
      span
        padding-left 10px
      a
        margin-left 10px
        text-decoration none
        color #fff
        vertical-align middle
      a:nth-of-type(2), :nth-of-type(3)
        display inline-block
        width 24px
        height 18px
        margin-bottom 5px
      a:nth-of-type(2)
        background url(./email.png) no-repeat center center / contain
      a:nth-of-type(3)
        background url(./email-red.png) no-repeat center center / contain
      a:nth-of-type(6)
        margin-right 10px
</style>
