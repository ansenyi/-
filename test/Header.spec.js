/* eslint-disable no-undef,no-unused-vars */
import 'babel-polyfill'
import { mount } from 'vue-test-utils'
import Header from '@/components/Header/Header.vue'
import Bus from '@/common/js/Bus'

describe('header', () => {
  function storageMock () {
    let storage = {}
    return {
      setItem: function (key, value) {
        storage[key] = value || ''
      },
      getItem: function (key) {
        return key in storage ? storage[key] : null
      },
      removeItem: function (key) {
        delete storage[key]
      },
      get length () {
        return Object.keys(storage).length
      },
      key: function (i) {
        let keys = Object.keys(storage)
        return keys[i] || null
      }
    }
  }
  // noinspection JSAnnotator
  window.localStorage = storageMock()

  let localStorage = storageMock()
  it('storage', () => {
    localStorage.setItem('a', 1)
    expect(localStorage.getItem('a')).toBe(1)
  })
  // window.localStorage.getItem = storageMock().getItem
  // window.localStorage.setItem = storageMock().setItem
  function initHeader (routeName) {
    const $route = {
      name: routeName
    }
    const cmp = mount(Header, {
      mocks: {
        $route
      }
    })
    return cmp
  }
  describe('placeholder or menu', () => {
    it('placeholder', () => {
      const cmp = initHeader('Login')
      console.log(cmp.vm.$route.name)
      expect(cmp.contains('#placeholder')).toBeTruthy()
      expect(cmp.contains('#menu')).toBeFalsy()
    })
    it('menu', () => {
      const cmp = initHeader('DevicePreview')
      expect(cmp.contains('#placeholder')).toBeFalsy()
      expect(cmp.contains('#menu')).toBeTruthy()
      expect(cmp.element).toMatchSnapshot()
    })
  })
  it('adminName', () => {
    window.localStorage.setItem('ADMIN_NAME', 'admin')
    const cmp = initHeader('DevicePreview')
    // cmp.update()
    cmp.vm.$nextTick(() => {
      expect(window.localStorage.getItem('ADMIN_NAME')).toBe('admin')
      expect(cmp.vm.adminName).toBe('admin')
      console.log('admin name')
    })
  })
  it('has message', () => {
    const cmp = initHeader('DevicePreview')
    expect(cmp.find('.email').hasStyle('display', 'none')).toBeFalsy()
    Bus.$emit('hasMessage', true)
    // cmp.update()
    // expect(cmp.find('.email-red').hasStyle('display', 'none')).toBeFalsy()
    cmp.vm.$nextTick(() => {
      expect(cmp.find('.email-red').hasStyle('display', 'none')).toBeFalsy()
    })
  })
  it('clear token', () => {
    window.localStorage.setItem('JWT_TOKEN', 'token')
    const cmp = initHeader('DevicePreview')
    cmp.vm.clearToken()
    cmp.vm.$nextTick(() => {
      expect(window.localStorage.getItem('JWT_TOKEN')).toBe(null)
    })
  })
  it('clear name', () => {
    window.localStorage.setItem('ADMIN_NAME', 'admin')
    const cmp = initHeader('DevicePreview')
    cmp.vm.clearName()
    cmp.vm.$nextTick(() => {
      expect(window.localStorage.getItem('ADMIN_NAME')).toBe(null)
      console.log('clear name')
    })
  })
})
