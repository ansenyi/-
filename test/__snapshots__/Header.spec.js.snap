// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`header placeholder or menu menu 1`] = `
<header>
  <div
    class="logo"
  >
    <router-link
      to="/"
    />
  </div>
    
  <nav
    id="menu"
  >
    <router-link
      to="/"
    >
      主菜单
    </router-link>
     
    <span>
      
    </span>
     
    <router-link
      class="email"
      to="/message"
    />
     
    <router-link
      class="email-red"
      style="display: none;"
      to="/message"
    />
     
    <router-link
      to="/upload"
    >
      上传软件
    </router-link>
     |
    
    <a
      class="logout"
      href="javascript:;"
    >
      退出
    </a>
     
    <a
      href=""
    >
      帮助
    </a>
  </nav>
</header>
`;
